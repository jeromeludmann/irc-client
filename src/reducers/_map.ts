import { Action } from "redux";

export const mapReducer = <S, E = void>(map: {
  [action: string]: (state: S, action: Action, extraStates?: E) => S;
}) => (state: S, action: Action, extraStates?: E): S =>
  map.hasOwnProperty(action.type)
    ? map[action.type](state, action, extraStates)
    : state;
