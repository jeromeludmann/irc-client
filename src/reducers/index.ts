import {
  RouteState,
  routeInitialState,
  reduceRoute,
} from "@app/reducers/route";
import {
  ServersState,
  serversInitialState,
  reduceServers,
} from "@app/reducers/servers";
import { RoutedAction } from "@app/Route";
import { Reducer } from "redux";

export interface AppState {
  readonly servers: ServersState;
  readonly route: RouteState;
}

export const rootInitialState = {
  servers: serversInitialState,
  route: routeInitialState,
};

export const reduce: Reducer<AppState, RoutedAction> = (
  state = rootInitialState,
  action,
) => ({
  servers: reduceServers(state.servers, action, { route: state.route }),
  route: reduceRoute(state.route, action, { servers: state.servers }),
});
