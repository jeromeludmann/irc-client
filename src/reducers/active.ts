import { Route, STATUS_WINDOW } from "@app/Route";
import { SwitchWindowAction, SWITCH_WINDOW } from "@app/actions/ui-window";
import { JoinAction, JOIN } from "@app/actions/message-in";
import { UserState } from "@app/reducers/user";

export type ActiveWindowState = Route;

export type ActiveWindowAction = JoinAction | SwitchWindowAction;

export const activeInitialState: ActiveWindowState = {
  server: "serverKey", // TODO multi servers
  window: STATUS_WINDOW,
};

type ExtraParams = {
  user: UserState;
};

export default (
  active = activeInitialState,
  action: ActiveWindowAction,
  { user }: ExtraParams,
): ActiveWindowState => {
  switch (action.type) {
    case JOIN:
      return action.payload.user.nick === user.nick ? action.route : active;

    case SWITCH_WINDOW:
      return action.route;

    default:
      return active;
  }
};
