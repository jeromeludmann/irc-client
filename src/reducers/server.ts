import { Action } from "redux";
import reduceModes, {
  ModesState,
  modesInitialState,
} from "@app/reducers/server-modes";
import { WindowState } from "@app/reducers/window";
import reduceBufferRouter, {
  BufferRouterState,
  bufferRouterInitialState,
  BufferRouterAction,
} from "@app/reducers/server-buffers";
import { Route } from "@app/Route";
import { UserState } from "@app/reducers/user";

export type ServerState = {
  readonly modes: ModesState;
  readonly buffers: BufferRouterState;
};

export type ServerAction = Action;

interface ExtraParams {
  user: UserState;
  route: Route;
  active: WindowState;
}

export const serverInitialState: ServerState = {
  modes: modesInitialState,
  buffers: bufferRouterInitialState,
};

export default (
  server = serverInitialState,
  action: ServerAction,
  params: ExtraParams,
): ServerState => ({
  modes: reduceModes(server.modes),
  buffers: reduceBufferRouter(
    server.buffers,
    action as BufferRouterAction,
    params,
  ),
});
