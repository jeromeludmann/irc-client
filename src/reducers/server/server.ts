import reduceModes, {
  ModesState,
  modesInitialState,
} from "@app/reducers/server/modes";
import { WindowState } from "@app/reducers/window";
import reduceBufferRouter, {
  BufferRouterState,
  bufferRouterInitialState,
  BufferRouterAction,
} from "@app/reducers/server/buffer-router";
import { Route } from "@app/Route";
import { UserState } from "@app/reducers/user";
import { RPL_MYINFO, IncomingReplyMyInfoAction } from "@app/actions/incoming";
import { Action } from "redux";

export type ServerState = {
  readonly name: string; // TODO
  readonly availableModes: { user: string[]; channel: string[] }; // TODO
  readonly modes: ModesState;
  readonly buffers: BufferRouterState;
};

export type ServerAction = Action;

interface ExtraParams {
  user: UserState;
  route: Route;
  active: WindowState;
}

export const serverInitialState: ServerState = {
  name: "<unknown>", // TODO
  availableModes: { user: [], channel: [] }, // TODO
  modes: modesInitialState,
  buffers: bufferRouterInitialState,
};

// TODO extract me
function reduceName(name: string, action: IncomingReplyMyInfoAction) {
  switch (action.type) {
    case RPL_MYINFO:
      return action.payload.serverName;
    default:
      return name;
  }
}

// TODO extract me
function reduceAvailableModes(
  availableModes: any,
  action: IncomingReplyMyInfoAction,
) {
  switch (action.type) {
    case RPL_MYINFO:
      return {
        user: action.payload.availableUserModes,
        channel: action.payload.availableChannelModes,
      };
    default:
      return availableModes;
  }
}

export default (
  server = serverInitialState,
  action: ServerAction,
  params: ExtraParams,
): ServerState => ({
  name: reduceName(server.name, action as IncomingReplyMyInfoAction),
  availableModes: reduceAvailableModes(
    server.availableModes,
    action as IncomingReplyMyInfoAction,
  ),
  modes: reduceModes(server.modes),
  buffers: reduceBufferRouter(
    server.buffers,
    action as BufferRouterAction,
    params,
  ),
});
