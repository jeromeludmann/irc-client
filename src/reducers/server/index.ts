import {
  ServerModesState,
  serverModesInitialState,
  reduceServerModes,
} from "@app/reducers/server/modes";
import { RouteState } from "@app/reducers/route";
import {
  ChannelsState,
  channelsInitialState,
  reduceChannels,
} from "@app/reducers/server/channels";
import { RoutedAction } from "@app/Route";
import {
  serverNameInitialState,
  reduceServerName,
} from "@app/reducers/server/name";
import {
  reduceAvailableServerModes,
  AvailableServerModesState,
  availableServerModesInitialState,
} from "@app/reducers/server/availableModes";
import { ServerNameState } from "@app/reducers/server/name";
import {
  UserState,
  userInitialState,
  reduceUser,
} from "@app/reducers/server/user";

export type ServerState = {
  readonly name: ServerNameState;
  readonly user: UserState;
  readonly availableModes: AvailableServerModesState;
  readonly modes: ServerModesState;
  readonly channels: ChannelsState;
};

export const serverInitialState: ServerState = {
  name: serverNameInitialState,
  user: userInitialState,
  availableModes: availableServerModesInitialState,
  modes: serverModesInitialState,
  channels: channelsInitialState,
};

export const reduceServer = (
  server = serverInitialState,
  action: RoutedAction,
  extraStates: { route: RouteState },
): ServerState => ({
  name: reduceServerName(server.name, action),
  user: reduceUser(server.user, action),
  availableModes: reduceAvailableServerModes(server.availableModes, action),
  modes: reduceServerModes(server.modes),
  channels: reduceChannels(server.channels, action, {
    ...extraStates,
    user: server.user,
  }),
});
