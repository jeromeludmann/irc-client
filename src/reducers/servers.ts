import { ServerState, reduceServer } from "@app/reducers/server";
import { RoutedAction } from "@app/Route";
import { RouteState } from "@app/reducers/route";

export interface ServersState {
  readonly [key: string]: ServerState;
}

export const serversInitialState: ServersState = {};

export const reduceServers = (
  servers = serversInitialState,
  action: RoutedAction,
  states: { active: RouteState },
): ServersState =>
  action.route && action.route.serverKey
    ? {
        ...servers,
        [action.route.serverKey]: reduceServer(
          servers[action.route.serverKey],
          action,
          states,
        ),
      }
    : servers;
