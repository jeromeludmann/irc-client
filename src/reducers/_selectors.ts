import { createSelector } from "reselect";
import { RootState } from "@app/reducers";
import { ServersState } from "@app/reducers/servers";
import { RouteState } from "@app/reducers/route";

export const selectServers = ({ servers }: RootState): ServersState => servers;

export const selectRoute = ({ route }: RootState): RouteState => route;

export const selectServer = createSelector(
  selectServers,
  selectRoute,
  (servers, { serverKey }) => servers[serverKey],
);
