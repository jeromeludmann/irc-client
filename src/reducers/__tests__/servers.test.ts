import { reduceServers, serversInitialState } from "@app/reducers/servers";
import { closeWindow } from "@app/actions/ui";
import { STATUS } from "@app/Route";
import { serverInitialState } from "@app/reducers/server";
import { lookup } from "@app/actions/socket";

describe("reduce servers", () => {
  test("without state", () => {
    expect(
      reduceServers(
        undefined,
        { type: "", route: { serverKey: "server1", channelKey: "" } },
        { route: { serverKey: "server1", channelKey: STATUS } },
      ),
    ).toMatchSnapshot();
  });

  describe("by dispatching close window", () => {
    test("that is not a status window", () => {
      expect(
        reduceServers(
          serversInitialState,
          closeWindow({ serverKey: "server1", channelKey: "#channel" }),
          { route: { serverKey: "server1", channelKey: "#channel" } },
        ),
      );
    });

    describe("that is a status window", () => {
      test("without server", () => {
        expect(
          reduceServers(
            serversInitialState,
            closeWindow({ serverKey: "server1", channelKey: STATUS }),
            { route: { serverKey: "server1", channelKey: STATUS } },
          ),
        );
      });

      test("with two servers", () => {
        expect(
          reduceServers(
            {
              server1: serverInitialState,
              server2: serverInitialState,
            },
            closeWindow({ serverKey: "server1", channelKey: STATUS }),
            { route: { serverKey: "server1", channelKey: STATUS } },
          ),
        );
      });
    });
  });

  describe("by dispatching lookup", () => {
    expect(
      reduceServers(
        { server1: serverInitialState },
        lookup("server1", null, "address", "family", "host"),
        { route: { serverKey: "server1", channelKey: STATUS } },
      ),
    );
  });
});
