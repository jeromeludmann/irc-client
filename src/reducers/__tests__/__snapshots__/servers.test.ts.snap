// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduce servers without state 1`] = `Object {}`;
